File: Clifford/clifford.cpp
/*
xn+1 = sin(a yn) + c cos(a xn)
yn+1 = sin(b xn) + d cos(b yn)
*/

#include "stdafx.h"

#include "clifford.h"

using namespace attractors;

void clifford::render(
	int quote, int pool,
	image& img,
	color::filter hue,
	std::function<void()> notify) const {

	color* bmp = img.data();
	int nx = img.nx, ny = img.ny;
	int fpc = frames / 100;

	number dX = nx / (maxX - minX), dY = ny / (maxY - minY);

	int i1 = quote * frames / pool, i2 = (quote + 1) * frames / pool;
	for (int i = i1; i < i2; i++) {
		const number p = static_cast<number>(i) / frames;
		const number a = cos(minA + p * (maxA - minA)) * 2.0;
		const number b = cos(minB + p * (maxB - minB)) * 2.0;
		const number c = cos(minC + p * (maxC - minC)) * 2.0;
		const number d = cos(minD + p * (maxD - minD)) * 2.0;

		const color curcol = hue(p);

		number x = 0.0, y = 0.0;
		number xn, yn;

		for (int j = 0; j < iters; j++) {
			xn = sin(a * y) + c * cos(a * x);
			yn = sin(b * x) + d * cos(b * y);
			x = xn;
			y = yn;

			if (j < skip)
				continue;

			int xi = static_cast<int>((x - minX) * dX);
			int yi = static_cast<int>((y - minY) * dY);
			if (xi >= 0 && xi < nx && yi >= 0 && yi < ny) {
				// there is no significant threads access to one data point at one time
				bmp[xi + yi * nx] += curcol;
			}
		}

		if (!(i % fpc)) {
			notify();
		}
	}
}

// The End.

File: Clifford/clifford.lua
-- starts

local filemask = "img%04d.tga"
local pool = 4 -- working threads pool size
local nx, ny = 1600, 1200 -- image dimension
local sensitivity = 0.02

-- xn+1 = sin(a yn) + c cos(a xn)
-- yn+1 = sin(b xn) + d cos(b yn)

local geomlst = {
--	{mina,    maxa,     minb,     maxb,     minc,     maxc,     mind,     maxd      },
	{ 1.6,     1.3,     -0.6,      1.7,     -1.2,      0.5,      1.6,      1.4,     },
	{ 0.6,     1.2,      1.2,      1.8,     -0.5,      1.8,      1.6,      1.4,     },
	{ 0.25,    1.75,    -1.12,     1.6,     -0.4,      1.8,     -1.8,      0.4,     },
	{ 1.6,     1.6,     -0.8,     -0.4,     -1.3,     -1.1,      1.5,      1.7,     },
}
local g = clifford.new() -- make clifford geometry structure
g.frames, g.iters, g.skip = 10000, 10000, 10 -- iterations parameters
g.minx, g.maxx = -4.0, 4.0
g.miny = g.minx * ny / nx
g.maxy = g.maxx * ny / nx

local img = image.create(nx, ny)
for i, p in ipairs(geomlst) do
	print(string.format("a[%g,%g] b[%g,%g] c[%g,%g] d[%g,%g]",
		p[1], p[2], p[3], p[4], p[5], p[6], p[7], p[8]))
	g.mina, g.maxa = math.acos(p[1] / 2.0), math.acos(p[2] / 2.0)
	g.minb, g.maxb = math.acos(p[3] / 2.0), math.acos(p[4] / 2.0)
	g.minc, g.maxc = math.acos(p[5] / 2.0), math.acos(p[6] / 2.0)
	g.mind, g.maxd = math.acos(p[7] / 2.0), math.acos(p[8] / 2.0)

	img:clear()
	local dur = g:render(pool, img, "hue") -- call multithreaded render
	print("\r"..dur.."s            ")

	local filename = string.format(filemask, i)
	io.write("writing...")
	img:writetga(filename, sensitivity)
	print("\r"..filename.."      ")
end

-- the end.
